Fix `/share` and cleanup and reorganize frontend locale loading

renchap opened the pull request:
My react-intl upgrade PR broke the locale loading on `/share`. This was caused by `getLocale` being called when importing files, as it is called in the root scope of the file.

This is a bad practice and can easily lead to those files being imported/loaded while the locale is not loaded yet. The current code needs to dynamically import the various React components after `loadLocale` has been called, which is error-prone.

Now, locale data is loaded when the component/code is called, not imported. This avoid the need to dynamically import this code once the locale data has been loaded

While I was working on this, I made some additional changes that I did not wanted to do in my initial PR to not make it too complex:
- Removed some unused files I missed in my last commit
- No more locale loading as a side-effect
- No need for the `locale` prop anymore for root React components, this is read from `<html lang>`
- Converted locale-related files to Typescript
- locale-related TS code is in `mastodon/locales/`, which only exports the needed functions & types

renchap committed:
Removed unused locale-data

renchap committed:
Rework locale loading:

- No need for the `locale` prop anymore for root React components, this is read from `<html lang>`
- Converted locale-related files to Typescript
- No more locale loading as a side-effect. Locale data is loaded when the component/code is called, not imported. This avoid the need to dynamically import this code once the locale data has been loaded

renchap commented:
Note: I can also rework this to not move the locale JSON files and have the TS code in another directory if you prefer (`mastodon/intl`?)

nschonni commented on .prettierignore:
```diff
@@ -61,7 +61,7 @@ docker-compose.override.yml
 
 # Ignore locale files
-/app/javascript/mastodon/locales
+/app/javascript/mastodon/locales/data/*.json
```
It seems like the JSON files don't need to be ignored anymore after the migration from `translationRunner.js`

renchap replied:
They are automatically generated, so we dont want Prettier to process them

renchap commented:
> I understand the reason for moving the JSON files and I agree with it, but I'm not sure how well Crowdin handles that, so I'd like that to be postponed until we make sure this won't break anything with Crowdin.

I can leave the JSON in `locales/` and move the TS files elsewhere, but I am not sure where. Ideas:
- `mastodon/intl/`
- `mastodon/locale/` (without an `s`)
- `mastodon/utils/intl/`

ClearlyClaire commented:
Do the TS files need to be kept out of that folder as a first step? If not, I'd postpone that split to a later PR, if yes, I'm fine with `mastodon/intl`

renchap committed:
Move all frontend locale-related code to `locales/`

`index.ts` re-exports what can be consumed by the app and nothing else

renchap commented:
I removed the JSON file renaming.

All TS files are in `mastodon/locales/` with the JSON files in the same directory. I split them in multiple files, `index.ts` re-exports what should be used in the app.
